### PR TITLE
Fix Bug 1227250, replace all instances of sendto donate links with donate.

### PR DESCRIPTION
**Pull request by schalkneethling**
@alexgibson @flodolo r?

### CODE REVIEW

**Comment by schalkneethling** [comment]
Just waiting for confirmation the leaving a default amount of 50 for all locales/currencies is ok:
https://bugzilla.mozilla.org/show_bug.cgi?id=1227250#c18

File: bedrock/settings/base.py [hunk 1]
```diff
@@ -596,31 +596,31 @@ def facebook_tab_url_lazy():
...
 
-DONATE_SPANISH_LINK = ('https://donate.mozilla.org/es/?presets=100,50,25,15&amount=50'
+DONATE_SPANISH_LINK = ('https://donate.mozilla.org/es/?amount=50'
     '&ref=EOYFR2015&utm_campaign=EOYFR2015&utm_source=mozilla.org'
     '&utm_medium=referral&utm_content={source}&currency=eur')
 
 DONATE_LOCALE_LINK = {
     'default': (
-        'https://sendto.mozilla.org/page/contribute/Give-Now?source={source}'
+        'https://donate.mozilla.org/en-US/?source={source}'
```
Comment (flodolo) [inline bedrock/settings/base.py]: IMO the default should not have en-US

**Comment by flodolo** [comment]
My personal feeling is that the DONATE_LOCALE_LINK thing is not the right approach (btw, it fails for Italian, since it doesn't link to https://donate.mozilla.org/it/ and €).

What's the difference between nn-NO and nb-no? Only the locale code.

So, it would make more sense to have a mapping like
```
DONATE_LOCALE_LINK = {
    'default': {
        'amount': 50,
        'currency': 'eur',
    },
   'en-US': {
        'amount': 50,
        'currency': 'usd',
    },
   'nb-NO': {
        'amount': 50,
        'currency': 'nok',
    },
...
```

And then use the `donate_url` helper function to build the link, including the current LOCALE in the URL.

File: bedrock/settings/base.py (hunk 1 above)
Comment (schalkneethling) [inline bedrock/settings/base.py]: I actually agree and was thinking that, because when they do get the locale switching implemented we will not be forcing en-US with the default. Because there was a note added to always add a locale I thought it best to leave it in. With that said, it will currently fallback to en-US anyhow so.... I reckon dropping that off is fine.

**Comment by schalkneethling** [comment]
The reason `it` fails to load for the correct locale is because we do not have an entry for it :-/ Which I found strange indeed. I do agree though but, perhaps we need to rethink/rework this mechanism in a separate bug?

Although I am open to discuss it here as well.

**Comment by flodolo** [comment]
TBH I think this is the right bug to fix it, no point in replacing a broken link with another broken link.

**Comment by flodolo** [comment]
Also note that random locales (or not available) will fallback to en-US
https://donate.mozilla.org/klingon/

**Comment by schalkneethling** [comment]
@flodolo Fair enough, this change will be needed for the takeover work also I reckon. Let me have a look at this and make additional changes.

**Comment by ScottDowne** [comment]
@flodolo 

Currently, random or not known locale fallback to the first browser set locale, if that's not known, it falls back to en-US.

We have an open bug to have this fallback to the next browser set locale, until we find a known one.

**Comment by flodolo** [comment]
@ScottDowne That was my point. 

IMO right now the right thing to do is pass a locale code explicitly: if we simply use donate.mozilla.org we fail for too many locales (Italian, all Spanish variants). If we pass the wrong locale, we still fallback to en-US, so we're good.

When locale detection is fixed on the other side, we can think about removing the locale code and let d.m.o do its work.

File: bedrock/mozorg/tests/test_helper_misc.py [hunk 2]
```diff
@@ -24,15 +24,15 @@
 
 TEST_DONATE_LOCALE_LINK = {
     'default':
-        'https://sendto.mozilla.org/page/contribute/Give-Now?source={source}',
+        'https://donate.mozilla.org/en-US/?source={source}',
     'en-US':
-        'https://sendto.mozilla.org/page/contribute/givenow-seq?'
-        'preset=2&source={source}&ref=EOYFR2014&utm_campaign=EOYFR2014'
-        '&utm_source=mozilla.org&utm_medium=referral&utm_content={source}',
+        'https://donate.mozilla.org/?amount=50'
```
Comment (ScottDowne) [inline bedrock/mozorg/tests/test_helper_misc.py]: Should we be explicit with en-US here, and not rely on defaults where we can help it?

Comment (schalkneethling) [inline bedrock/mozorg/tests/test_helper_misc.py]: We can be explicit. @flodolo also suggested that we rethink the way we build these URLs and perhaps use something more like https://github.com/mozilla/donate.mozilla.org/blob/master/data/currencies.js

Basically we should be able to inject more than just the `{source}`, ideally we should also do `{locale}` and `{currency}'. Perhaps we can also set thing like presets and defaults but, then we probably want to always set them unless, passing a parameter such as `presets=` would be fine on the donate.m.o side.

Comment (ScottDowne) [inline bedrock/mozorg/tests/test_helper_misc.py]: Yeah, I think I agree. If you did something like that, we could get Andrea to give us a list of slightly different amounts and presets for mozilla.org, which might perform better.

I would also suggest to set the default amount to be the same as the second (middle button) in the preset.

**Comment by schalkneethling** [comment]
@flodolo Ok, before I get to far into adding all the locales in `base.py` you wanna have a look at what I have there already? Also, I have some questions such as, which countries should all use Euro and which uses GBP?

Let me know your thoughts, thanks!

File: bedrock/mozorg/helpers/misc.py [hunk 3]
```diff
@@ -349,11 +349,11 @@ def press_blog_url(ctx):
...
 def donate_url(ctx, source=''):
-    """Output a link to the donation page taking locales into account.
+    """Output a link to the donation page formatted using settings.DONATE
 
-    Uses the locale from the current request. Checks to see if we have
-    a donation page that match this locale, returns the localized page
-    url or falls back to the US page url if not.
+    Formats the donation url using custom values defined in settings.DONATE for
+    the curent locale. If a locale is not defined, it falls back to default, except
+    for the locale, which is always the locale from the request.
```
Comment (flodolo) [inline bedrock/mozorg/helpers/misc.py]: Typo: current. Kind of hard to follow the second paragraph (do you really need the part after except)

File: bedrock/mozorg/helpers/misc.py [hunk 4]
```diff
@@ -365,33 +365,31 @@ def donate_url(ctx, source=''):
...
-    if locale not in settings.DONATE_LOCALE_LINK:
-        if 'default' in settings.DONATE_LOCALE_LINK:
-            locale = 'default'
-        else:
-            locale = 'en-US'
+    index = locale
+
+    # if the current locale is not available in settings.DONATE,
+    # we want to fallback to the default params but, still pass
+    # the current locale in the URL
```
Comment (flodolo) [inline bedrock/mozorg/helpers/misc.py]: Maybe add a note why we're doing it (waiting for donate.mozilla.org to fix the issue with locale detection).

File: bedrock/settings/base.py [hunk 5]
```diff
@@ -1,3 +1,6 @@
+#!/usr/bin/env python
```
Comment (flodolo) [inline bedrock/settings/base.py]: Do you need the shebang here? It's not an executable file

File: bedrock/settings/base.py (hunk 1 above)
Comment (flodolo) [inline bedrock/settings/base.py]: I guess you don't need this anymore, right?

**Comment by flodolo** [comment]
Makes sense to me, even if I don't think we need the currency symbol?

Check https://github.com/mozilla/donate.mozilla.org/blob/master/data/locale-data.js for currency+locale (GBP is only en-GB)

**Comment by schalkneethling** [comment]
Thx @flodolo - I was thinking the currency symbol could be useful for the front page takeover but, perhaps we do not really need it there? We could pass it to the template from the view based on locale.

**Comment by flodolo** [comment]
You would know how to write the currency symbol, not how to write the string: before or after the number? With a space or without? Normal or non-breaking?

File: bedrock/mozorg/helpers/misc.py (hunk 3 above)
Comment (schalkneethling) [inline bedrock/mozorg/helpers/misc.py]: That is what happens when someone talks to you while you think and type ;p I will update it to make more sense.

File: bedrock/settings/base.py (hunk 1 above)
Comment (schalkneethling) [inline bedrock/settings/base.py]: Nope, I was leaving the removal until the end.

File: bedrock/settings/base.py (hunk 5 above)
Comment (schalkneethling) [inline bedrock/settings/base.py]: It is for the currency symbols. When running the tests it fails. Seeing that having the symbols here is not going to be of much use, I am going to remove them and then these lines as well.

**Comment by schalkneethling** [comment]
@flodolo Indeed, it is a lot more complex than just knowing the symbol. The current implementation we are using for the takeover will probably work best then.

**Comment by schalkneethling** [comment]
@flodolo @ScottDowne this is ready for review.

File: bedrock/mozorg/helpers/misc.py [hunk 6]
```diff
@@ -365,33 +361,33 @@ def donate_url(ctx, source=''):
...
-            locale = 'default'
-        else:
-            locale = 'en-US'
+    index = locale
+
+    # if the current locale is not available in settings.DONATE,
+    # we want to fallback to the default params but, still pass
+    # the current locale in the URL. This is a workaround whilst
+    # donate.m.o does not handle automaic locale negotiation.
+    # see https://bugzilla.mozilla.org/show_bug.cgi?id=1227199#c5
```
Comment (flodolo) [inline bedrock/mozorg/helpers/misc.py]:
```
# If the current locale is not available in settings.DONATE,
# fallback to the default params but still pass the current 
# locale in the URL. This is a temporary workaround since 
# donate.mozilla.org does not fully handle automatic
# locale negotiation.
```

File: bedrock/settings/base.py [hunk 7]
```diff
@@ -596,93 +596,587 @@ def facebook_tab_url_lazy():
...
-        'https://sendto.mozilla.org/page/content/paypal-donate-thb/?'
-        'source={source}&ref=EOYFR2014&utm_campaign=EOYFR2014'
-        '&utm_source=mozilla.org&utm_medium=referral&utm_content=PPcurrency_THB'
-    ),
+DONATE_LINK = ('https://donate.mozilla.org/{locale}/?presets={presets}'
+    '&amount={default}&ref=EOYFR2015&utm_campaign=EOYFR2015'
+    '&utm_source=mozilla.org&utm_medium=referral&utm_content={source}'
+    '&currency={currency}')
+
+DONATE_PARAMS = {
```
Comment (flodolo) [inline bedrock/settings/base.py]: Please double check the list of locales with the locales actually supported on Bedrock, no point adding locales that will never be available on mozilla.org. Where did you get the current list?

From a very quick check (there might be others):
* af-ZA -> af
* au (not even a locale, it's a region, and we don't support en-AU anyway)
* ca-ES -> ca
* ch (again, not a locale)
* en-CA
* en-NZ
* en-PH
* et-EE -> et
* eu-ES
* fa-IR
* fl and fl-FL. Not even sure what this is… Did you mean Finnish (fi)?
* fr-FR, fr-CA
* fy
* hr-HR
* hu-HU
* id-ID
* ig-NG
* is-IS
* no
* xh-ZA

Comment (schalkneethling) [inline bedrock/settings/base.py]: Thanks @flodolo I am relying on your knowledge here. I have gone overboard on some I realize.

**Comment by schalkneethling** [comment]
@flodolo ok, The locales are now only those in prod locales. How does this look?

**Comment by flodolo** [comment]
To be completely honest I'm still not sure how you built this list.

The starting point should be these
https://github.com/mozilla/donate.mozilla.org/blob/master/data/currencies.js
https://github.com/mozilla/donate.mozilla.org/blob/master/data/locale-data.js

You should drop all the currencies that are not available on d.m.o

Drop: ach, af, az, be, bg, br, bs, ee, en-ZA, es-AR, es-CL, hr, hy-AM, ig, is, ka, kzt, km, ko, ln, mai, mk, my, ro, si, sq, sw, uk, ur, uz, vi, xh, yo, zh-CN, zh-TW, zu.

bn-BD and bn-IN should be 'in'

But please double check with the two files in d.m.o, given the size I probably missed some.

File: bedrock/settings/base.py (hunk 7 above)
Comment (schalkneethling) [inline bedrock/settings/base.py]: How does the list look to you now @flodolo? Thanks!

File: bedrock/settings/base.py [hunk 8]
```diff
@@ -596,93 +596,322 @@ def facebook_tab_url_lazy():
...
+        'currency': 'eur',
+        'presets': '100,50,25,15',
+        'default': '50'
+    },
+    'en-GB': {
+        'currency': 'gbp',
+        'presets': '150,100,50,25',
+        'default': '100'
+    },
+    'en-CA': {
```
Comment (flodolo) [inline bedrock/settings/base.py]: Remove this, not supported on Bedrock

File: bedrock/settings/base.py [hunk 9]
```diff
@@ -596,93 +596,322 @@ def facebook_tab_url_lazy():
...
+        'currency': 'cad',
+        'presets': '150,100,50,25',
+        'default': '100'
+    },
+    'en-US': {
+        'currency': 'usd',
+        'presets': '100,50,25,15',
+        'default': '50'
+    },
+    'es': {
```
Comment (flodolo) [inline bedrock/settings/base.py]: No need for this, we redirect es to es-ES https://www.mozilla.org/es/

File: bedrock/settings/base.py [hunk 10]
```diff
@@ -596,93 +596,322 @@ def facebook_tab_url_lazy():
...
+        'currency': 'inr',
+        'presets': '5000,3000,2000,1000',
+        'default': '3000'
+    },
+    'sk': {
+        'currency': 'eur',
+        'presets': '100,50,25,15',
+        'default': '50'
+    },
+    'si': {
```
Comment (flodolo) [inline bedrock/settings/base.py]: This is wrong. Did you mean 'sl' (Slovenian)? 'si' is Sinhala (India)

**Comment by flodolo** [comment]
Minus a couple of nits, the list looks good to me (maybe ask someone from MoFo to double check if they're around?). Also, spot-checked it with de, it, sv-SE and everything works as expected.

File: bedrock/settings/base.py (hunk 10 above)
Comment (schalkneethling) [inline bedrock/settings/base.py]: Indeed I did

File: bedrock/settings/base.py (hunk 7 above)
Comment (schalkneethling) [inline bedrock/settings/base.py]: @ScottDowne want to give this a look over and see if there is anything you would like to change/remove

**Comment by ScottDowne** [comment]
Any locales we don't currently support on d.m.o we can add.

They would initially be in English, but then we could at least add currency and presets.

**Comment by schalkneethling** [comment]
@flodolo Thoughts?

**Comment by flodolo** [comment]
I don't see any advantage in adding all locales, since we would just fall back to USD for the currency. 

We're passing the locale code nevertheless in the URL, even if a locale is not available in this array; only the params fallback to en-US.

**Comment by schalkneethling** [comment]
@ScottDowne Does that work for you? If so, I reckon this is good to merge.

**Comment by flodolo** [comment]
Note: mine doesn't really count as a proper code review, I'd suggest to have a second pair of eyes on it.

**Comment by schalkneethling** [comment]
@pmclanahan r?

**Comment by schalkneethling** [comment]
Thanks for all the help @flodolo and @ScottDowne

File: bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html [hunk 11]
```diff
@@ -183,11 +183,11 @@ <h4 class="section-title" id="drumbeat">
       <p>
         {% trans
           logo=static('img/foundation/annualreport/2009/drumbeat-logo.jpg'),
-          webcraft="http://www.drumbeat.org/",
-          festival="http://www.drumbeat.org/",
-          movies="http://www.drumbeat.org/",
-          subtitles="http://www.drumbeat.org/",
-          fund="https://donate.mozilla.org/"
+          webcraft="//www.drumbeat.org/webcraft",
```
Comment (pmclanahan) [inline bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html]: Why not specifically use "https://" links here instead of protocol relative? Means that it will be "http://" during local dev and could mask issues.

File: bedrock/mozorg/helpers/misc.py [hunk 12]
```diff
@@ -365,33 +361,33 @@ def donate_url(ctx, source=''):
...
+    if locale not in settings.DONATE_PARAMS:
```
Comment (pmclanahan) [inline bedrock/mozorg/helpers/misc.py]: NIT: Could just be:

```python
donate_url_params = settings.DONATE_PARAMS.get(locale, settings.DONATE_PARAMS['default'])
```

File: bedrock/settings/base.py [hunk 13]
```diff
@@ -596,93 +596,312 @@ def facebook_tab_url_lazy():
...
-        'source={source}&ref=EOYFR2014&utm_campaign=EOYFR2014'
-        '&utm_source=mozilla.org&utm_medium=referral&utm_content=PPcurrency_THB'
-    ),
+DONATE_LINK = ('https://donate.mozilla.org/{locale}/?presets={presets}'
+    '&amount={default}&ref=EOYFR2015&utm_campaign=EOYFR2015'
+    '&utm_source=mozilla.org&utm_medium=referral&utm_content={source}'
+    '&currency={currency}')
+
+DONATE_PARAMS = {
+    'default': {
```
Comment (pmclanahan) [inline bedrock/settings/base.py]: This is mostly a nit, but I'd find it easier to understand and more explicit if this entry were `en-US` instead of `default`, then have the default for a locale that's not in the list be `en-US`. That way the code is clear that we're defaulting to en-US. Also it means that the majority of the traffic will go down the main code path (finding the locale in the dict) instead of the alternate branch (using the default).

**Comment by pmclanahan** [comment]
Looks pretty good to me. Just a couple of comments I'm interested in your feedback on. 

r+wc

File: bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html (hunk 11 above)
Comment (schalkneethling) [inline bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html]: @alexgibson Has resolved these issues in a separate PR [https://github.com/mozilla/bedrock/pull/3608], the only change here is to use the `donate_url` helper for the `fund` variable

File: bedrock/mozorg/helpers/misc.py (hunk 12 above)
Comment (schalkneethling) [inline bedrock/mozorg/helpers/misc.py]: Thx @pmclanahan - If we remove `default` though, I guess we won't need the above at all

File: bedrock/settings/base.py (hunk 13 above)
Comment (schalkneethling) [inline bedrock/settings/base.py]: I completely agree. The reason I had a `default` is two-fold. One, there was a default before (not a great reason) and two, I thought they might want to set the presets and default differently for people who fallback to `en-US`. I don't believe that is going to be the case so, I am going to make this change.

File: bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html (hunk 11 above)
Comment (alexgibson) [inline bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html]: It looks like you've accidentally reverted the changes when rebasing. This needs fixing again.

**Comment by schalkneethling** [comment]
@pmclanahan updated, thanks!

File: bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html (hunk 11 above)
Comment (schalkneethling) [inline bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html]: Bah! Yes, there was a merge conflict and I wanted to keep my change to use the `donate_url` and did not notice I copied over that one as well. Fixing. Apologies for that @alexgibson

File: bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html [hunk 14]
```diff
@@ -183,11 +183,11 @@ <h4 class="section-title" id="drumbeat">
       <p>
         {% trans
           logo=static('img/foundation/annualreport/2009/drumbeat-logo.jpg'),
-          webcraft="http://www.drumbeat.org/",
-          festival="http://www.drumbeat.org/",
-          movies="http://www.drumbeat.org/",
-          subtitles="http://www.drumbeat.org/",
-          fund="https://donate.mozilla.org/"
+          webcraft="http://www.drumbeat.org/webcraft",
```
Comment (alexgibson) [inline bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html]: You're still reverting changes made in the other PR here. The only thing that needs to change is the donate link.

Comment (schalkneethling) [inline bedrock/foundation/templates/foundation/annualreport/2009/broadening-our-scope.html]: :anguished: Fixing

**Comment by alexgibson** [comment]
r+ from me (seeing as @pmclanahan already gave an r+wc, and we need this in master for the takeover PR), looks good. :hamburger:

File: bedrock/mozorg/helpers/misc.py (hunk 12 above)
Comment (pmclanahan) [inline bedrock/mozorg/helpers/misc.py]: We do need the above for locales not in the list.

Comment (pmclanahan) [inline bedrock/mozorg/helpers/misc.py]: We'd just need to change it to: 

```python
donate_url_params = settings.DONATE_PARAMS.get(locale, settings.DONATE_PARAMS['en-US'])
```

Comment (schalkneethling) [inline bedrock/mozorg/helpers/misc.py]: Ah, I thought it would just fallback to en-US. Will open a PR to fix that.

Comment (schalkneethling) [inline bedrock/mozorg/helpers/misc.py]: Oh wait, I have this:
https://github.com/mozilla/bedrock/blob/master/bedrock/mozorg/helpers/misc.py#L377

Comment (pmclanahan) [inline bedrock/mozorg/helpers/misc.py]: Oh yeah. That's perfect. All good :)
